### DISCUSSION TITLE
Wie bekomme ich den Datenpunkt 2404_BivalenceControlMode in den Klartext

### DISCUSSION THREAD

**Discussion by elli75** [MyHomeMyData/E3onCAN#7] [Q&A]
Hallo, der Datenpunkt " e3oncan.0.HPMUMASTER_0x680.tree.2404_BivalenceControlMode " sollte 3 Temperaturen ausgeben die bei einer Vitocal 250-SH Hybrid Wärmepumpe die Freigabe des externen wärmeerzeugers steuert.
Aktuell ist der Anzeigewert : "033200ecff01"
Was muss ich tun um hier die Temperaturen angezeigt und über ein Script auch ändern zu können?

Danke schon einmal

**Comment by MyHomeMyData** [comment]
Hallo @elli75 ,
Temperaturen werden meistens in 16-Bit Integer mit Vorzeichen und Umrechnungsfaktor 10 übertragen. Um auszuprobieren, ob das bei diesem Datenpunkt auch so ist, kannst Du folgendes tun:

Editiere den State `e3oncan.0.HPMUMASTER_0x680.info.udsDidsSpecific` und füge **nach der ersten öffnenden Klammer** folgendes ein:
```
  "2404": {
    "codec": "O3EComplexType",
    "len": 6,
    "id": "BivalenceControlMode",
    "args": {
      "subTypes": [
        {
          "codec": "O3EInt16",
          "len": 2,
          "id": "t1",
          "args": {
            "scale": 10,
            "signed": true,
            "offset": 0
          }
        },
        {
          "codec": "O3EInt16",
          "len": 2,
          "id": "t2",
          "args": {
            "scale": 10,
            "signed": true,
            "offset": 0
          }
        },
        {
          "codec": "O3EInt16",
          "len": 2,
          "id": "t3",
          "args": {
            "scale": 10,
            "signed": true,
            "offset": 0
          }
        }
      ]
    }
  },
```
Nun speicherst Du den State, dabei das Häkchen bei "Bestätigt" **nicht** setzen.

Wenn Du jetzt den Datenpunkt 2404 nochmal abrufst, sollten drei Temperaturen t1, t2 und t3 angezeigt werden. Falls die Werte sinnvoll sind, kannst Du auch versuchen, einen Wert zu ändern. Dazu einfach den eine der Temperaturen editieren und speichern (wieder "Bestätigt" **nicht** anhaken). Jetzt versucht der Adapter, den Wert zu schreiben. Nach 2,5 Sekunden wird der Datenpunkt gelesen und siehst, ob die Änderung übernommen wurde. Am besten auch mal ins Log schauen.

Wenn das funktioniert und Du die Temperaturen zuordnen kannst, sollten wir für t1,t2,t3 sinnvolle (englische) Bezeichnungen finden. Ich würde das dann in die Gesamtliste der Datenpunkte übernehmen und mit dem nächsten Update bekommen alle diese Info.

**Comment by elli75** [comment]
Super das einfügen hat schon einmal geklappt. 2404 erscheint jetzt als ausklappbares Ordnersymbol mit den 3 Werten t1,t2 undt3.
Die angezeigten Werte sind jedoch nicht plausibel.
Hinter 2404 verbergen sich 4 Unterwerte. 
2404.0 " Betriebsprogramm " mit einstellbaren Wertebereich von 0,1,2,und3
2404.1 " Bivalenter Temperaturpunkt " mit einstellbaren Wertebereich von -50°c bis +50°C
2404.2 " Alternativer Temperaturpunkt "   mit einstellbaren Wertebereich von -50°c bis +50°C
2404.3 " Regelungsmodi " mit einstellbaren Wertebereich von 0,1,2,und3

Hier müsste der " State e3oncan.0.HPMUMASTER_0x680.info.udsDidsSpecific " noch angepasst werden.

**Comment by MyHomeMyData** [comment]
Dann versuch es mal mit der Struktur unten. Ich habe erstmal die deutschen Bezeichnungen übernommen.
Bitte den vorhandenen Block "2404" durch den neuen ersetzen. Am besten auch den Datenpunkt `e3oncan.0.HPMUMASTER_0x680.tree.2404_BivalenceControlMode` komplett löschen. Wird beim ersten Abrufen den Datenpunktes neu angelegt.
Gibt es die Beschreibung auch auf englisch? Dann könnten wir die Originalbezeichnungen verwenden. Kannst Du einen Link zur Beschreibung posten?

```
  "2404": {
    "codec": "O3EComplexType",
    "len": 6,
    "id": "BivalenceControlMode",
    "args": {
      "subTypes": [
        {
          "codec": "O3EByteVal",
          "len": 1,
          "id": "0_Betriebsprogramm",
          "args": {
            "offset": 0
          }
        },
        {
          "codec": "O3EInt16",
          "len": 2,
          "id": "1_BivalenterTemperaturpunkt",
          "args": {
            "scale": 10,
            "signed": true,
            "offset": 0
          }
        },
        {
          "codec": "O3EInt16",
          "len": 2,
          "id": "2_AlternativerTemperaturpunkt",
          "args": {
            "scale": 10,
            "signed": true,
            "offset": 0
          }
        },
        {
          "codec": "O3EByteVal",
          "len": 1,
          "id": "3_Regelungsmodus",
          "args": {
            "offset": 0
          }
        }
      ]
    }
  },
```

**Comment by elli75** [comment]
Damit geht es, die 4 Datenpunkte sind wie gewünscht angelegt und die Werte sind richtig. , siehe Bild
![2024-04-25_190300](https://github.com/MyHomeMyData/E3onCAN/assets/68502497/ffe49507-0e48-4713-b14c-a6ed863935d3)

Die Werte lassen sich aber nicht schreiben, bzw der Wert spring 2 Sekunden nach dem ändern wieder auf den letzten Wert zurück.
Das Attribut ist aber  als "writeable" gesetzt.

**Comment by elli75** [comment]
![2024-04-25_190849](https://github.com/MyHomeMyData/E3onCAN/assets/68502497/ba7bc52f-3ba3-4c08-8a41-3c097c49652a)

**Reply by MyHomeMyData** [reply]
Dieses "Writeable" bedeuted ja nur, dass Du den State innerhalb von ioBroker ändern kannst. Ob der Datenpunkt im Viessmann-Gerät verändert werden kann, legt Viessmann fest. Für 2404 sollte das zwar der Fall sein, es kann aber sein, dass man dazu eine höhere Berechtigung benötigt. Was steht denn im ioBroker-Log, wenn Du versuchst, einen Wert zu ändern?

**Comment by elli75** [comment]
nter " e3oncan.0.HPMUMASTER_0x680.info.udsDidsSpecific " eintagen um den Datenpunkt " e3oncan.0.HPMUMASTER_0x680.tree.2831_BivalenceControlAlternativeTemperature " sichtbar zu machen? Aktuell steht da  der Wert " ecff "

**Comment by elli75** [comment]
Da ist etwas schief gelaufen. Im Log vom IoBrker steht:
![2024-04-25_193100](https://github.com/MyHomeMyData/E3onCAN/assets/68502497/3e3a4e8d-1053-4cd3-b915-41ff84bf8dd7)

Ich kann die Werte aber an der Regelung der Wärmepumpe ändern ( nach Eingabe des Passwortes für den Heizungsbauer )
![IMG_20240425_144204](https://github.com/MyHomeMyData/E3onCAN/assets/68502497/da530794-b62e-4a29-8ef6-ffd30bbbd01c)


Was muss ich unter " e3oncan.0.HPMUMASTER_0x680.info.udsDidsSpecific " eintagen um den Datenpunkt " e3oncan.0.HPMUMASTER_0x680.tree.2831_BivalenceControlAlternativeTemperature " sichtbar zu machen? Aktuell steht da der Wert " ecff "

**Reply by MyHomeMyData** [reply]
Ok, das habe ich vermutet. 2404 ist ein Datenpunkt, der besonders geschützt ist (Seed&Key-Verfahren). Der kann über den Adapter nicht geändert werden. Die Anlage schickt dann eine "negative response".
Außerdem ist der Umrechnungsfaktor falsch. Im Display steht z.B. -20, im Datenpunkt -2. Der Umrechnungsfaktor ist also 1, nicht 10. Das müsstest Du in der Datenpunktbeschreibung für die Temperaturen ändern in `"scale": 1,`.

Zu 2831: Das müsste dann wohl so aussehen:
```
  "2831": {
    "codec": "O3EInt16",
    "len": 2,
    "id": "BivalenceControlAlternativeTemperature",
    "args": {
      "scale": 1,
      "signed": true,
      "offset": 0
    }
  },
```

Auch hier gilt: Den Datenpunkt `e3oncan.0.HPMUMASTER_0x680.tree.2831_BivalenceControlAlternativeTemperature` einmal löschen.

**Comment by elli75** [comment]
Nein der Umrechnungsfaktorist schon richtig. Bei Viessmann muss man bei vielen Werten "durch 10 teilen".
-20 sind -2,0°C und 50 ist 5,0°C

**Reply by MyHomeMyData** [reply]
Hätte ich mir denken können, ist ja ein Viessmann Gerät ...

Dann muss vermutlich bei "BivalenceControlAlternativeTemperature" auch `"scale": 10,` stehen.

**Reply by MyHomeMyData** [reply]
@elli75 Habe die Datenpunkte 2404 und 2831 in `ioBroker.e3oncan` eingepflegt, Branch [develop](https://github.com/MyHomeMyData/ioBroker.e3oncan/tree/develop). Kommt dann mit dem nächsten Release automatisch in die `udsDidsCommon` rein. Wenn die Datenpunkte auch in der `udsDidsSpecific` eingetragen sind, müssen sie nach dem Update dort gelöscht werden, um die Definitionen aus der Common-Liste wirksam zu machen.

**Comment by elli75** [comment]
Hallo, ich brauch nochmal Expertenhilfe. Dieses Mal geht es um den Datenpunkt 2225. Ich vermute dass sich darüber die aktuelle Lade / Entladeleistung in / aus dem Stromspeicher abrufen läßt.

![2225](https://github.com/MyHomeMyData/E3onCAN/assets/68502497/69b380e8-5c60-4148-ae8f-815bb2fcd859)

 ich habe es mit dieser Änderung in  udsDidsSpecific  versucht aber damit geht es nicht.

### "2225": {
    "codec": "RawCodec",
    "len": 12,
    "id": "ElectricEnergyStorageSetpoint",
    "args": {
      "subTypes": [
        {
          "codec": "O3EByteVal",
          "len": 1,
          "id": "0_Ladeleistung",
          "args": {
            "offset": 0
          }
        },
        {
          "codec": "O3EInt16",
          "len": 2,
          "id": "1_Entladeleistung",
          "args": {
            "scale": 10,
            "signed": true,
            "offset": 0
          }
        },
        {
          "codec": "O3EInt16",
          "len": 2,
          "id": "2_LeistungLebensdauer",
          "args": {
            "scale": 10,
            "signed": true,
            "offset": 0
          }
        }
      ]
    }
  },

Was muss ich ändern um den Datenpunkt auslesen zu können.

**Reply by MyHomeMyData** [reply]
@elli75 Bei meiner VX3 ist die  Info zur Lade-/Entlade-Leistung im Datenpunkt `1836_ElectricalEnergyStorageCurrentPower` verfügbar.

Bei `2225_ElectricEnergyStorageSetpoint` dürfte es sich eher um einen oder mehrere Sollwerte handeln, oder? Bei mir liefert der Datenpunkt im Moment `000000000000000000000000`, die Batterie ist allerdings auch grade inaktiv (100% geladen, PV speist ein).

Deine Datenpunktedefinition für 2225 ist nicht konsistent. Was erwartest Du denn für eine Struktur in dem Datenpunkt? Hast Du dazu Infos oder eine Unterlage? Schick mir das oder Deine Idee unstrukturiert, dann baue ich daraus ein passendes JSON. Gerne auch die Roh-Daten bei bestimmten Betriebszuständen (am besten mit zugehörigen Zahlenwerten aus ViCare).

**Reply by elli75** [reply]
Ja stimmt, jetzt wo die Sonne weg ist  und der Speicher entladen wird kommen auch keine Interessanten Daten von 2225.
Hauptsächlich bin ich auf der Suche nach einer Möglichkeit den Verbrauch des Hauses angezeit zu bekommen.
Erzeugung minus Einspeisung  (+/- Batterieladung /entladung )hat ja noch die Unsicherheit des Eigenverbrauchs des Wechselrichter und der Wandlungverluste.

**Reply by MyHomeMyData** [reply]
@elli75 Den Verbrauch des Haushalts berechne ich so:
`1603_PointOfCommonCouplingPower.ActivePower + 1690_ElectricalEnergySystemPhotovoltaicStatus.ActivePower_cumulated + 1836_ElectricalEnergyStorageCurrentPower`

**Comment by elli75** [comment]
Hallo, kannst Du im nächsten Update des Adapters den Datenpunkt" 1799_PrimaryCircuitCurrentTemperatureSetpoint " einpflegen?
Die 1799 gibt die aktuelle Solltemperatur der Wärmepumpe an. Im iobroker wird der Wert aktuell als "0000" angezeigt. Hier sollte im Klartext z.B. 28,5 °C stehen.

Danke schon einmal

**Reply by MyHomeMyData** [reply]
Übernehme ich mit dem nächsten Update. Du kannst das direkt einpflegen, wenn Du in `udsDidsSpecific` folgendes ergänzt:
```
  "1799": {
    "codec": "O3EInt16",
    "len": 2,
    "id": "PrimaryCircuitCurrentTemperatureSetpoint",
    "args": {
      "scale": 10,
      "signed": true,
      "offset": 0
    }
  },
```
Möglichst vorher den Datenpunkt 1799 im `tree` Objektbaum löschen, damit das Zahlenformat nach der Änderung stimmt. Und beim Speichern von `udsDidsSpecific` bei `Ack` das Häkchen weglassen.

**Comment by elli75** [comment]
Danke, hat funktioniert.

**Comment by elli75** [comment]
Hallo, ich hab mal wieder einen Datenpunkt, der noch eingepflegt werden könnte.

e3oncan.0.HPMUMASTER_0x680.tree.2741_ComfortEnsuringMode

Hier drüber kann der Notbetrieb aktiviert werden.
Wäre nett, wenn Du den hinzufügen könntest.
Gruß Elli
